python/private/py_info.bzl

```diff
@@ -113,3 +116,124 @@ This field is currently unused in Bazel and may go away in the future.
+        set_has_py2_only_sources = lambda *a, **k: _PyInfoBuilder_set_has_py2_only_sources(self, *a, **k),
+        set_has_py3_only_sources = lambda *a, **k: _PyInfoBuilder_set_has_py3_only_sources(self, *a, **k),
+        set_uses_shared_libraries = lambda *a, **k: _PyInfoBuilder_set_uses_shared_libraries(self, *a, **k),
+        transitive_pyc_files = builders.DepsetBuilder(),
+        transitive_sources = builders.DepsetBuilder(),
+    )
+    return self
+
+def _PyInfoBuilder_get_has_py3_only_sources(self):
+    return self._has_py3_only_sources[0]
+
+def _PyInfoBuilder_get_has_py2_only_sources(self):
+    return self._has_py2_only_sources[0]
+
+def _PyInfoBuilder_set_has_py2_only_sources(self, value):
+    self._has_py2_only_sources[0] = value
+    return self
+
+def _PyInfoBuilder_set_has_py3_only_sources(self, value):
+    self._has_py3_only_sources[0] = value
+    return self
+
+def _PyInfoBuilder_merge_has_py2_only_sources(self, value):
+    self._has_py2_only_sources[0] = self._has_py2_only_sources[0] or value
+    return self
+
+def _PyInfoBuilder_merge_has_py3_only_sources(self, value):
+    self._has_py3_only_sources[0] = self._has_py3_only_sources[0] or value
+    return self
+
+def _PyInfoBuilder_merge_uses_shared_libraries(self, value):
+    self._uses_shared_libraries[0] = self._uses_shared_libraries[0] or value
+    return self
+
+def _PyInfoBuilder_get_uses_shared_libraries(self):
+    return self._uses_shared_libraries[0]
+
+def _PyInfoBuilder_set_uses_shared_libraries(self, value):
+    self._uses_shared_libraries[0] = value
+    return self
+
+def _PyInfoBuilder_merge(self, *infos):
+    return self.merge_all(infos)
+
+def _PyInfoBuilder_merge_all(self, py_infos):
+    for info in py_infos:
+        self.imports.add(info.imports)
+        self.merge_has_py2_only_sources(info.has_py2_only_sources)
+        self.merge_has_py3_only_sources(info.has_py3_only_sources)
+        self.merge_uses_shared_libraries(info.uses_shared_libraries)
+        self.transitive_sources.add(info.transitive_sources)
+
+        # BuiltinPyInfo doesn't have these fields
+        if hasattr(info, "transitive_pyc_files"):
+            self.transitive_pyc_files.add(info.transitive_pyc_files)
+
+    return self
+
+def _PyInfoBuilder_merge_target(self, target):
+    if PyInfo in target:
+        self.merge(target[PyInfo])
+    elif BuiltinPyInfo in target:
+        self.merge(target[BuiltinPyInfo])
+    return self
+
+def _PyInfoBuilder_merge_targets(self, targets):
+    for t in targets:
+        self.merge_target(t)
+    return self
+
+def _PyInfoBuilder_build(self):
+    if config.enable_pystar:
+        kwargs = dict(
+            direct_pyc_files = self.direct_pyc_files.build(),
+            transitive_pyc_files = self.transitive_pyc_files.build(),
+        )
+    else:
+        kwargs = {}
+
+    # todo: for bazel 6, make this return BuiltinPyInfo
```

Comment:
in 2 months this will be an outdated comment. So it is fine if we never get to this. Is it blocking the merging of the PR? will it break users on bazel 6?

Reply:
Oh oops, outdated comment. The _EffectivePyInfo trick addressed this. Removed.